File: commander_service.py
"""
Commander Randomizer Service Layer
This module provides high-level functions for commander selection.
Can be used by any frontend (GUI, Web, CLI, API).
"""

import random_commander
import scryfall_api


class CommanderService:
    """Service class for commander operations."""
    
    def __init__(self):
        self.csv_files = {
            'Weekly': 'top_commanders_week.csv',
            'Monthly': 'top_commanders_month.csv',
            '2-Year': 'top_commanders_2year.csv'
        }
        self.scryfall = None
        try:
            self.scryfall = scryfall_api.ScryfallAPI()
        except:
            pass  # Scryfall API optional
    
    def get_available_time_periods(self):
        """Get list of available time periods."""
        return list(self.csv_files.keys())
    
    def get_csv_info(self):
        """
        Get information about all CSV files.
        
        Returns:
            dict: {period: {'file': filename, 'max_rank': int, 'count': int}}
        """
        info = {}
        for period, csv_file in self.csv_files.items():
            try:
                commanders = random_commander.load_commanders(csv_file)
                max_rank = max(c['rank'] for c in commanders) if commanders else 0
                info[period] = {
                    'file': csv_file,
                    'max_rank': max_rank,
                    'count': len(commanders)
                }
            except Exception as e:
                info[period] = {
                    'file': csv_file,
                    'max_rank': 0,
                    'count': 0,
                    'error': str(e)
                }
        return info
    
    def randomize(self, 
                  time_period='Monthly',
                  min_rank=1,
                  max_rank=300,
                  quantity=3,
                  colors=None,
                  color_mode='exactly',
                  num_colors=None,
                  exclude_partners=False):
        """
        Randomize commanders with given parameters.
        
        Args:
            time_period: One of 'Weekly', 'Monthly', '2-Year'
            min_rank: Minimum rank (1+)
            max_rank: Maximum rank
            quantity: Number of commanders to select
            colors: Comma-separated color string (e.g., 'W,U,B') or None
            color_mode: 'exactly', 'including', or 'atmost'
            num_colors: Exact number of colors to filter by (0 for colorless) or None
            exclude_partners: Whether to exclude partner commanders
            
        Returns:
            dict: {
                'success': bool,
                'commanders': list of commander dicts,
                'total_available': int,
                'filter_description': str,
                'error': str (if success=False)
            }
        """
        try:
            # Get CSV file for time period
            csv_file = self.csv_files.get(time_period)
            if not csv_file:
                return {
                    'success': False,
                    'error': f'Invalid time period: {time_period}'
                }
            
            # Load commanders
            commanders = random_commander.load_commanders(csv_file)
            total_loaded = len(commanders)
            
            # Filter out partners if requested
            if exclude_partners:
                commanders = [c for c in commanders if ' // ' not in c['name']]
            
            # Build filter description
            filter_desc = f"{time_period} ranks {min_rank}-{max_rank}"
            if colors is not None or num_colors is not None:
                if num_colors is not None:
                    filter_desc += f" with exactly {num_colors} color(s)"
                    if colors:
                        mode_desc = {'exactly': 'exactly', 'including': 'including', 'atmost': 'at most'}
                        filter_desc += f" ({mode_desc[color_mode]}: {colors})"
                elif colors:
                    mode_desc = {'exactly': 'exactly', 'including': 'including', 'atmost': 'at most'}
                    filter_desc += f" with {mode_desc[color_mode]} colors: {colors}"
                elif colors == '':
                    filter_desc += " (colorless only)"
            
            if exclude_partners:
                filter_desc += " (excluding partners)"
            
            # Select random commanders
            selected = random_commander.select_random_commanders(
                commanders, min_rank, max_rank, quantity, colors, color_mode, num_colors
            )
            
            return {
                'success': True,
                'commanders': selected,
                'total_available': len(commanders),
                'total_loaded': total_loaded,
                'filter_description': filter_desc,
                'quantity_requested': quantity,
                'quantity_returned': len(selected)
            }
            
        except FileNotFoundError as e:
            return {
                'success': False,
                'error': f'CSV file not found: {csv_file}'
            }
        except Exception as e:
            return {
                'success': False,
                'error': str(e)
            }
    
    def get_commander_url(self, commander_name):
        """Get EDHREC URL for a commander."""
        return random_commander.commander_name_to_url(commander_name)
    
    def get_card_image(self, commander_name, version='normal'):
        """
        Get card image from Scryfall.
        
        Args:
            commander_name: Name of the commander
            version: Image version ('small', 'normal', 'large', 'png', etc.)
            
        Returns:
            PIL.Image or None
        """
        if not self.scryfall:
            return None
        
        try:
            return self.scryfall.get_card_image(commander_name, version)
        except Exception as e:
            print(f"Error fetching image for {commander_name}: {e}")
            return None
    
    def get_card_image_url(self, commander_name, version='normal'):
        """
        Get card image URL from Scryfall (without downloading).
        
        Args:
            commander_name: Name of the commander
            version: Image version ('small', 'normal', 'large', 'png', etc.)
            
        Returns:
            str: Image URL or None
        """
        if not self.scryfall:
            return None
        
        try:
            return self.scryfall.get_card_image_url(commander_name, version)
        except Exception as e:
            print(f"Error fetching image URL for {commander_name}: {e}")
            return None


# Singleton instance for convenience
_service_instance = None


def get_service():
    """Get singleton CommanderService instance."""
    global _service_instance
    if _service_instance is None:
        _service_instance = CommanderService()
    return _service_instance

File: docs/js/services/commanderService.js
// ========================================
// COMMANDER SERVICE
// ========================================

import { CSV_FILES } from '../config.js';
import { loadCSV } from '../dataLoader.js';
import { selectRandomCommanders, selectRandomCommandersWeighted } from '../commanderFilters.js';
import { commanderNameToUrl } from '../api/edhrec.js';
import { getCardImageUrl } from '../api/scryfall.js';
import { displayTextResults, displayCardImagesProgressive, clearResults, updateStatus, getResultMessage } from '../ui/display.js';
import { validateInputs, validateColorConfiguration } from '../ui/validation.js';
import { getColorFilterSettings, getAdditionalFilterSettings } from '../ui/events.js';
import { saveLastResults } from '../storage.js';
import { getDistributionFunction } from '../advancedRandomizer.js';

export async function randomizeCommanders(timePeriod, minRank, maxRank, quantity, colors, colorMode, numColors, selectedColorCounts, excludePartners, minCmc, maxCmc, saltMode, distributionFunc = null) {
    try {
        const csvFile = CSV_FILES[timePeriod];
        if (!csvFile) {
            return {
                success: false,
                error: `Invalid time period: ${timePeriod}`
            };
        }
        
        // Load commanders
        let commanders = await loadCSV(csvFile);
        const totalLoaded = commanders.length;
        
        // Filter out partners if requested
        if (excludePartners) {
            commanders = commanders.filter(c => !c.name.includes(' // '));
        }
        
        // Build filter description
        let filterDesc = `${timePeriod} ranks ${minRank}-${maxRank}`;
        if (colors !== null || numColors !== null || (selectedColorCounts && selectedColorCounts.length > 0)) {
            if (numColors !== null) {
                filterDesc += ` with exactly ${numColors} color(s)`;
                if (colors) {
                    const modeDesc = {'exactly': 'exactly', 'including': 'including', 'atmost': 'at most'}[colorMode];
                    filterDesc += ` (${modeDesc}: ${colors})`;
                }
            } else if (selectedColorCounts && selectedColorCounts.length > 0) {
                // Multi-select color counts
                const countsStr = selectedColorCounts.sort((a, b) => a - b).join(', ');
                filterDesc += ` with ${countsStr} color(s)`;
                
                if (colors) {
                    const modeDesc = {'exactly': 'exactly', 'including': 'including', 'atmost': 'at most'}[colorMode];
                    filterDesc += ` (${modeDesc}: ${colors})`;
                }
            } else if (colors) {
                const modeDesc = {'exactly': 'exactly', 'including': 'including', 'atmost': 'at most'}[colorMode];
                filterDesc += ` with ${modeDesc} colors: ${colors}`;
            } else if (colors === '') {
                filterDesc += ' (colorless only)';
            }
        }
        
        if (excludePartners) {
            filterDesc += ' (excluding partners)';
        }
        
        // Add CMC filter to description
        if (minCmc !== null && maxCmc !== null) {
            filterDesc += `, CMC ${minCmc}-${maxCmc}`;
        }
        
        // Add salt filter to description
        if (saltMode === 'salty') {
            filterDesc += `, Salty commanders`;
        } else if (saltMode === 'chill') {
            filterDesc += `, Chill commanders`;
        }
        
        // Add distribution mode to description
        if (distributionFunc) {
            filterDesc += ` (Advanced Randomizer)`;
        }
        
        // Select random commanders using weighted or uniform distribution
        const selected = distributionFunc 
            ? selectRandomCommandersWeighted(
                commanders, minRank, maxRank, quantity, colors, colorMode, numColors, selectedColorCounts, minCmc, maxCmc, saltMode, distributionFunc
            )
            : selectRandomCommanders(
                commanders, minRank, maxRank, quantity, colors, colorMode, numColors, selectedColorCounts, minCmc, maxCmc, saltMode
            );
        
        return {
            success: true,
            commanders: selected,
            total_available: commanders.length,
            total_loaded: totalLoaded,
            filter_description: filterDesc,
            quantity_requested: quantity,
            quantity_returned: selected.length
        };
        
    } catch (error) {
        return {
            success: false,
            error: error.message
        };
    }
}

export async function handleRandomize() {
    console.log('handleRandomize called');
    
    // Validate inputs
    const validation = validateInputs();
    if (!validation) {
        console.error('Validation failed');
        return;
    }
    
    const { minRank, maxRank, quantity } = validation;
    const timePeriod = document.getElementById('time-period').value;
    const { colors, color_mode, num_colors, selected_color_counts } = getColorFilterSettings();
    const { enable_cmc, min_cmc, max_cmc, enable_salt, salt_mode } = getAdditionalFilterSettings();
    const excludePartners = document.getElementById('exclude-partners').checked;
    const useTextOutput = document.getElementById('text-output').checked;
    
    // Get the distribution function (returns null if not using advanced randomizer)
    const distributionFunc = getDistributionFunction();
    
    // Validate color configuration
    const colorValidation = validateColorConfiguration();
    
    console.log('Request params:', { minRank, maxRank, quantity, timePeriod, colors, color_mode, num_colors, selected_color_counts, enable_cmc, min_cmc, max_cmc, enable_salt, salt_mode });
    
    // Clear previous results but preserve sort state
    clearResults(true);
    
    // Show loading
    updateStatus('Loading commanders...');
    document.getElementById('randomize-btn').disabled = true;
    document.getElementById('loading-indicator').classList.remove('hidden');
    
    try {
        // Call service function
        const result = await randomizeCommanders(
            timePeriod,
            minRank,
            maxRank,
            quantity,
            colors,
            color_mode,
            num_colors,
            selected_color_counts,
            excludePartners,
            min_cmc,
            max_cmc,
            salt_mode,
            distributionFunc
        );
        
        console.log('Service response:', result);
        
        if (!result.success) {
            alert(`Error: ${result.error}`);
            updateStatus(`Error: ${result.error}`);
            return;
        }
        
        const commanders = result.commanders;
        const commandersCount = commanders ? commanders.length : 0;
        
        // Generate appropriate status message
        const statusMessage = getResultMessage(result, quantity, colorValidation);
        
        // Display results
        if (commandersCount > 0) {
            if (useTextOutput) {
                // For text output, still fetch all URLs first (needed for display)
                for (const commander of commanders) {
                    commander.edhrec_url = commanderNameToUrl(commander.name);
                    commander.image_url = await getCardImageUrl(commander.name);
                }
                displayTextResults(result);
            } else {
                // For card images, display immediately and load images progressively
                displayCardImagesProgressive(commanders);
            }
            updateStatus(statusMessage);
            
            // Save the results for next visit
            saveLastResults(commanders);
        } else {
            updateStatus(statusMessage);
        }
        
    } catch (error) {
        console.error('Error:', error);
        alert(`An error occurred: ${error.message}`);
        updateStatus(`Error: ${error.message}`);
    } finally {
        document.getElementById('randomize-btn').disabled = false;
        document.getElementById('loading-indicator').classList.add('hidden');
    }
}
